Assets/Scripts/ServiceManagers/ServiceManager.cs

```diff
@@ -0,0 +1,117 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Author: Paul Hazen (paul.hazen@playeveryware.com)
+ *
+ * Date:   2024-06-17
+ *
+ * Notes: This file contains implementation of functionality that is common to
+ *        many of the different managers. Not all managers make use of it - as
+ *        transitioning will be done gradually and as areas are refactored.
+ *        Consequently, only EOSAchievementManager and StatsManager make use of
+ *        this base class.
+ *
+ */
+
+//#define ENABLE_DEBUG_EOSSTATSMANAGER
+
+namespace PlayEveryWare.EpicOnlineServices
+{
+    using System;
+    using System.Collections.Generic;
+
+    using Epic.OnlineServices;
+    using Epic.OnlineServices.Connect;
+
+    /// <summary>
+    /// Contains implementation of common functionality between different
+    /// EOS Service managers (Currently EOSAchievementManager and StatsManager).
+    /// </summary>
+    public abstract class ServiceManager : IEOSOnConnectLogin, IEOSSubManager
+    {
+        /// <summary>
+        /// Stores a list of functions to be called whenever data related to
+        /// the manager has been updated.
+        /// </summary>
+        protected IList<Action> _updateCallbacks = new List<Action>();
+
+        /// <summary>
+        /// Called when connect login has taken place.
+        /// </summary>
+        /// <param name="loginCallbackInfo"></param>
+        public void OnConnectLogin(LoginCallbackInfo loginCallbackInfo)
+        {
+            if (loginCallbackInfo.ResultCode != Result.Success)
+            {
+                UnityEngine.Debug.Log($"Player login did not succeed. Result code: {Enum.GetName(typeof(Result), loginCallbackInfo.ResultCode)}");
+                return;
+            }
+            OnPlayerLogin(loginCallbackInfo.LocalUserId);
+        }
+
+        /// <summary>
+        /// Should be called when a player has been successfully logged in.
+        /// </summary>
+        /// <param name="productUserId">
+        /// The ProductUserId of the player that has just been logged in.
+        /// </param>
+        protected abstract void OnPlayerLogin(ProductUserId productUserId);
+
+        /// <summary>
+        /// Implementation of this function should refresh the locally cached
+        /// data managed by the service manager in question.
+        /// </summary>
+        public abstract void Refresh();
+
+        /// <summary>
+        /// Trigger the notification callbacks that have been registered,
+        /// letting any consumers know that the data base been updated.
+        /// </summary>
+        protected void NotifyUpdated()
```

Comment:
Is there ever a world where `_updateCallbacks` has a callback, and during `action()` being called it calls `RemoveUpdateCallback`, modifying this list? That'd break because of the foreach not liking its collection being modified while enumerating.

Reply:
Or alternatively while this is running, it somehow causes something to add to the list of callbacks. That'd also break it.

Reply:
Because `NotifyUpdated` is _not_ async, and because none of the functions that update it are asynchronous, the list will not (as far as I understand things) be updated from another thread.

Reply:
It doesn't have to be another thread. What if one of the callbacks itself results in modifying this list?

Reply:
Excellent point! So, I changed the implementation so that before iterating over the callback list, a copy of the list is made, and the copy is iterated over instead.